vertical-pod-autoscaler/pkg/utils/status/status_object_test.go

```diff
@@ -0,0 +1,264 @@
+				}
+
+				return true, nil, fmt.Errorf("unsupported action")
+			},
+			wantErr: false,
+		},
+		{
+			name: "getting status object - retryable error",
+			getReactor: func() func(action testcore.Action) (bool, runtime.Object, error) {
+				i := 0
+				return func(action testcore.Action) (bool, runtime.Object, error) {
+					if action.GetVerb() == "get" && action.GetResource().Resource == "leases" {
+						i++
+						switch i {
+						case 1, 2:
+							return true, nil, syscall.ECONNRESET
+						default:
+							return true, &apicoordinationv1.Lease{}, nil
+						}
+					}
+
+					return true, nil, fmt.Errorf("unsupported action")
+				}
+			}(),
+			wantErr: false,
+		},
+		{
+			name: "getting status object - non-retryable error",
+			getReactor: func() func(action testcore.Action) (bool, runtime.Object, error) {
+				i := 0
+				return func(action testcore.Action) (bool, runtime.Object, error) {
+					if action.GetVerb() == "get" && action.GetResource().Resource == "leases" {
+						i++
+						switch i {
+						case 1:
+							return true, nil, fmt.Errorf("non-retryable error")
+						default:
+							return true, &apicoordinationv1.Lease{}, nil
+						}
+					}
+
+					return true, nil, fmt.Errorf("unsupported action")
+				}
+			}(),
+			wantErr: true,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("test case: %s", tc.name), func(t *testing.T) {
+			fc := fake.NewSimpleClientset()
+			fc.PrependReactor("get", "leases", tc.getReactor)
+			client := NewClient(fc, leaseName, leaseNamespace, 10*time.Second, leaseName)
+			_, err := client.GetStatus()
+			assert.True(t, (err != nil) == tc.wantErr)
+		})
+	}
+}
+
+func TestIsStatusValid(t *testing.T) {
+	tests := []struct {
+		name          string
+		lease         *apicoordinationv1.Lease
+		leaseTimeout  time.Duration
+		expectedValid bool
+	}{
+		{
+			name: "Valid CreationTimestamp",
+			lease: &apicoordinationv1.Lease{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1.Time{Time: time.Now()},
+				},
+			},
+			leaseTimeout:  10 * time.Second,
+			expectedValid: true,
+		},
+		{
+			name: "Valid RenewTime",
+			lease: &apicoordinationv1.Lease{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1.Time{Time: time.Now().Add(-time.Minute)},
+				},
+				Spec: apicoordinationv1.LeaseSpec{
+					RenewTime: &metav1.MicroTime{Time: time.Now()},
+				},
+			},
+			leaseTimeout:  10 * time.Second,
+			expectedValid: true,
+		},
+		{
+			name: "Outdated CreationTimestamp and RenewTime",
+			lease: &apicoordinationv1.Lease{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1.Time{Time: time.Now().Add(-time.Minute)},
+				},
+				Spec: apicoordinationv1.LeaseSpec{
+					RenewTime: &metav1.MicroTime{Time: time.Now().Add(-time.Minute)},
+				},
+			},
+			leaseTimeout:  10 * time.Second,
+			expectedValid: false,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("test case: %s", tc.name), func(t *testing.T) {
```

Comment:
These test cases are indeterministic. I think you want an injectable time.Now() if you's like to test this properly.

Reply:
I added `innerClock` in status_object.go.
I'm not happy with package level variable, but I don't see another way to inject time to `IsStatusValid` method :/

Reply:
A simpler alternative is to have a helper function isStatusValid(lease, timeout, now)  test it instead and have IsStatusValid just call(isStatusValid, timeout, time.Now() )

GarbageCollection of container states does sth like this: https://github.com/kubernetes/autoscaler/blob/b232915c2485345d3ae8e587bc3fa933bc5c70d9/vertical-pod-autoscaler/pkg/recommender/model/cluster.go#L331

Reply:
Done.